pkg/cassandra/util.go

```diff
@@ -31,14 +33,38 @@ func DatacenterStopping(dc *cassdcapi.CassandraDatacenter) bool {
 	return dc.GetConditionStatus(cassdcapi.DatacenterStopped) == corev1.ConditionTrue && dc.Status.CassandraOperatorProgress == cassdcapi.ProgressUpdating
 }
 
-func ComputeSystemReplication(kluster *api.K8ssandraCluster) SystemReplication {
+// GetDatacentersForSystemReplication determines the DCs that should be included for
+// replication. This function should only be used for system keyspaces. Replication for
+// system keyspaces is initially set through the management-api, not CQL. This allows us
+// to specify non-existent DCs for replication even though Cassandra 4 does not allow that.
+// That cannot be done when configuration replication through CQL which is why this func
+// should only be used for system keyspaces.
+func GetDatacentersForSystemReplication(kc *api.K8ssandraCluster) []api.CassandraDatacenterTemplate {
+	var datacenters []api.CassandraDatacenterTemplate
+	if kc.Status.GetConditionStatus(api.CassandraInitialized) == corev1.ConditionTrue {
+		datacenters = make([]api.CassandraDatacenterTemplate, 0)
+		for _, dc := range kc.Spec.Cassandra.Datacenters {
+			if status, found := kc.Status.Datacenters[dc.Meta.Name]; found && status.Cassandra.GetConditionStatus(cassdcapi.DatacenterReady) == corev1.ConditionTrue {
+				datacenters = append(datacenters, dc)
+			}
+		}
+	} else {
+		datacenters = kc.Spec.Cassandra.Datacenters
+	}
+	return datacenters
+}
+
+func ComputeSystemReplication(kc *api.K8ssandraCluster) SystemReplication {
 	rf := 3.0
-	for _, dc := range kluster.Spec.Cassandra.Datacenters {
+
+	datacenters := GetDatacentersForSystemReplication(kc)
```

Comment:
Couldn't you simplify this to just `	datacenters := kc.Spec.Cassandra.Datacenters` ? Since we determined that this code is only used for computing the initial system replication?

Reply:
No we can't do that. It would result in DCs added to an existing cluster to be included which leads to incorrect value for initial system replication. I

Reply:
Hmm but this function is never going to be called if you add a dc to an existing cluster, since the existing cluster resource would already have its `SystemReplicationAnnotation` set to a previous, initial value, right?

Reply:
That is correct. The function is called from `K8ssandraClusterReconciler.checkInitialSystemReplication()`. The first thing it does is check for the annotation. If the annotation exists and can be unmarshalled, the function returns so `ComputeInitialSystemReplication()` and `GetDatacentersForSystemReplication()` won't get called.